---
title: "Evaluation of DE analysis results on real data"
date: "`r format(Sys.time(), '%Y/%m/%d')`"
output: html_notebook
params:
  output_dir: NULL
editor_options:
  chunk_output_type: console
---

## 1. Preparations

```{r setup, cache=FALSE}
library(RSQLite)

library(hydroGOF)
library(tximport)
library(edgeR)

library(ggpubr)
library(ggcorrplot)
library(ggedit)
library(UpSetR)
library(plotROC)

library(tidyverse)

utl <- new.env(); source(here::here("scripts/utils/base.R"), utl); source(here::here("scripts/utils/plot.R"), utl)
h.align <- new.env(); source(here::here("scripts/helpers/align_helper.R"), h.align)
h.quant <- new.env(); source(here::here("scripts/helpers/quant_helper.R"), h.quant)
h.de <- new.env(); source(here::here("scripts/helpers/de_helper.R"), h.de)
h.real <- new.env(); source(here::here("scripts/helpers/real_helper.R"), h.real)

PREFIX <- "real"

output_dir <- ifelse(length(params$output_dir) > 0, params$output_dir, "results/figs_tbls")
```

## 2. Load assets

### 2.1 Load annotations

```{r}
annotations <- utl$load_annotations()
```

### 2.2 Load qPCR result

```{r}
qpcr_dir <- here::here("assets/datasets/GSE5350")

refseq2gencode <-
  file.path(qpcr_dir, "refid2gencode.tsv") %>% data.table::fread(sep = ",", header = TRUE)

result_qpcr_long <- file.path(qpcr_dir, "MAQC_TAQ_merged.txt") %>%
  h.real$load_qpcr_result() %>%
  left_join(refseq2gencode, by = "ID") %>%
  group_by(gene_id, group) %>%
  summarize(mean(value))

result_qpcr <- left_join(
  result_qpcr_long %>% filter(group == "A"),
  result_qpcr_long %>% filter(group == "B"),
  by = "gene_id"
) %>%
  select(gene_id, `mean(value).x`, `mean(value).y`) %>%
  rename(ctrl = `mean(value).x`) %>%
  rename(case = `mean(value).y`) %>%
  rename(feature_id = gene_id)

result_qpcr <- result_qpcr %>% mutate(mean = mean(ctrl, case))
result_qpcr$logfc <- log2(result_qpcr$case / result_qpcr$ctrl)
```

### 2.3 Find files

```{r}
PATTERNS_TX <- c(
  cuffdiff = "de_cuffdiff/isoform_exp.diff$",
  ebseq = "de_ebseq/transcript/result.tsv$",
  ballgown = "de_ballgown/result_transcript.tsv$",
  sleuth = "de_sleuth/result_transcript_wt.tsv"
)

PATTERNS_GENE <- c(
  cuffdiff = "de_cuffdiff/gene_exp.diff$",
  ebseq = "de_ebseq/gene/result.tsv$",
  ballgown = "de_ballgown/result_gene.tsv$",
  sleuth = "de_sleuth/result_gene_wt.tsv"
)

.names <- function(p) {
  paste(
    names(p) %>% strsplit("\\.") %>% sapply("[", 1),
    p %>% map_chr( ~ .x %>% utl$to_combination()),
    sep = "."
  )
}

.exclude <- function(p) {
  .keep <- names(p) %>%
    grepl("gencode_basic", .) %>% !.

  .keep <- .keep &
    names(p) %>%
    grepl("gencode_refseq", .) %>% !.

  .keep <- .keep &
    (
      names(p) %>%
        grepl("tophat2-cuffdiff", .) |
        names(p) %>%
        grepl("hisat2-stringtie-ballgown", .) |
        names(p) %>%
        grepl("star-rsem-ebseq", .) |
        names(p) %>%
        grepl("kallisto-sleuth", .)
    )
  
  p[.keep]
}

# NOTE: For real comparison (subsampled)
input_dirs_sub <-
  here::here("results", c("05_00_sub", "05_01_sub", "05_02_sub"))

names(input_dirs_sub) <- c("00", "01", "02")

paths_de_real_sub <-
  input_dirs_sub %>% lapply(utl$find_paths, PATTERNS_TX) %>% unlist

names(paths_de_real_sub) <- .names(paths_de_real_sub)

paths_de_real_sub <- paths_de_real_sub %>% .exclude

# NOTE: For mock comparison (MAQCA)
input_dirs_maqca <-
  here::here("results",
             c("04_00_maqca_mock", "04_01_maqca_mock", "04_02_maqca_mock"))

names(input_dirs_maqca) <- c("00", "01", "02")

paths_de_real_maqca <-
  input_dirs_maqca %>% lapply(utl$find_paths, PATTERNS_TX) %>% unlist

names(paths_de_real_maqca) <- .names(paths_de_real_maqca)

paths_de_real_maqca <- paths_de_real_maqca %>% .exclude

# NOTE: For mock comparison (MAQCB)
input_dirs_maqcb <-
  here::here("results",
             c("06_00_maqcb_mock", "06_01_maqcb_mock", "06_02_maqcb_mock"))

names(input_dirs_maqcb) <- c("00", "01", "02")

paths_de_real_maqcb <-
  input_dirs_maqcb %>% lapply(utl$find_paths, PATTERNS_TX) %>% unlist

names(paths_de_real_maqcb) <- .names(paths_de_real_maqcb)

paths_de_real_maqcb <- paths_de_real_maqcb %>% .exclude

# NOTE: For comparison to qpcr result
input_dirs_all <- here::here("results", c("03_merge_ercc_frrf"))
names(input_dirs_all) <- c("00")

paths_de_real_all_tx <-
  input_dirs_all %>% lapply(utl$find_paths, PATTERNS_TX) %>% unlist

names(paths_de_real_all_tx) <- .names(paths_de_real_all_tx)

paths_de_real_all_tx <- paths_de_real_all_tx %>% .exclude

paths_de_real_all_gene <-
  input_dirs_all %>% lapply(utl$find_paths, PATTERNS_GENE) %>% unlist

names(paths_de_real_all_gene) <- .names(paths_de_real_all_gene)

paths_de_real_all_gene <- paths_de_real_all_gene %>% .exclude

# NOTE: For comparison to simulation
input_dirs_sim <- here::here("results", c("test01_main"))
names(input_dirs_sim) <- c("00")

paths_de_sim_all_gene <-
  input_dirs_sim %>% lapply(utl$find_paths, PATTERNS_GENE) %>% unlist

names(paths_de_sim_all_gene) <- .names(paths_de_sim_all_gene)

paths_de_sim_all_gene <- paths_de_sim_all_gene %>% .exclude
```

### 2.4 Load DE results

```{r}
results_de_real_sub <- paths_de_real_sub %>% h.real$load_results()

results_de_real_maqca <-paths_de_real_maqca %>% h.real$load_results()

results_de_real_maqcb <-paths_de_real_maqca %>% h.real$load_results()

results_de_real_all_tx <-
  paths_de_real_all_tx %>% h.real$load_results()

results_de_real_all_gene <-
  paths_de_real_all_gene %>% h.real$load_results()

results_de_sim_all_gene <-
  paths_de_sim_all_gene %>% h.real$load_results()
```

## 2. Compare with validation data

```{r}
path_annotation <- here::here("shared/assets/references/grch38/annotations/gencode/gencode.v31.annotation.gtf")

annotations <- utl$load_annotations()
annotation_main <- annotations$gencode

transcript_lengths <- path_annotation %>%
  utl$load_gtf(types = c("exon")) %>%
  group_by(transcript_id) %>%
  summarise(length = sum(end - start + 1)) %>%
  data.frame

# FIXME: Gene-level biotype
features <- bind_rows(
  annotation_main %>%
    distinct(transcript_id, transcript_name, biotype) %>%
    set_names(c("feature_id", "feature_name", "biotype")),
  annotation_main %>%
    distinct(gene_id, gene_name, gene_type) %>%
    set_names(c("feature_id", "feature_name", "biotype"))
)

paths_mappability <- here::here(
  "results/mappabilities/100/_work",
  c(
    "gencode.v31.transcripts.formatted.gencode.v31.transcripts.formatted.aligned.merged.sqlite"
  )
) %>% set_names(c("gencode"))

load_mappability <- function(path, features, feature_type) {
  conn <- RSQLite::SQLite() %>% RSQLite::dbConnect(path, synchronous = "off")

  df <- RSQLite::dbGetQuery(conn, paste0("select * from ", feature_type, "_mappabilities;")) %>%
    set_names(c("feature_id", "mappability")) %>%
    left_join(features, by = "feature_id")

  RSQLite::dbDisconnect(conn)

  df
}

feature_types <- c("gene", "transcript")

mappabilities <- paths_mappability %>%
  enframe(value = "path") %>%
  crossing(feature_type = feature_types) %>%
  mutate(data = map2(path, feature_type, ~ load_mappability(.x, features = features, .y)))

mappabilities1 <- mappabilities %>%
  unnest(data) %>%
  filter(mappability == 1) %>%
  group_by(name, path, feature_type) %>%
  nest

mappabilities <- mappabilities %>%
  unnest(data) %>%
  filter(mappability != 1) %>%
  group_by(name, path, feature_type) %>%
  nest

mappability_feature_ids <- mappabilities %>%
  filter(name == "gencode") %>%
  filter(feature_type == "gene") %>%
  unnest(data) %>%
  filter(feature_id %in% pull(!!result_qpcr, "feature_id")) %>%
  utl$set_bin("mappability", num_bins = 4) %>%
  select(feature_id, bin) %>%
  group_by(bin) %>%
  nest %>%
  ungroup %>%
  mutate(feature_ids = map(data, ~ pull(.x, feature_id))) %>%
  select(-data) %>%
  arrange(bin)

mappability_feature_ids <- mappability_feature_ids %>%
  bind_rows(
    tibble(
      bin = NA,
      feature_id = mappabilities1 %>%
        unnest(data) %>%
        filter(name == "gencode") %>%
        filter(feature_type == "gene") %>%
        filter(feature_id %in% pull(!!result_qpcr, "feature_id")) %>%
        pull(feature_id)
    ) %>% group_by(bin) %>%
      nest %>%
      mutate(feature_ids = map(data, ~ pull(.x, feature_id))) %>%
      select(-data) %>%
      arrange(bin)
  )
  
mean_abundance_feature_ids <- result_qpcr %>%
  select(feature_id, mean) %>%
  utl$set_bin("mean", num_bins = 4, exclude = TRUE) %>%
  select(bin, feature_id) %>%
  group_by(bin) %>%
  nest %>%
  ungroup %>%
  mutate(feature_ids = map(data, ~ pull(.x, feature_id))) %>%
  select(-data) %>%
  arrange(bin)

mappability_mean_abundance_feature_ids <- left_join(
  mappability_feature_ids %>%
    unnest(feature_ids),
  mean_abundance_feature_ids %>%
    unnest(feature_ids),
  by = "feature_ids") %>%
  dplyr::rename(feature_id = "feature_ids") %>%
  relocate(feature_id) %>%
  arrange(bin.x, bin.y)

mappability_mean_abundance_feature_ids <- mappability_mean_abundance_feature_ids %>%
  group_by(bin.x, bin.y) %>%
  nest %>%
  mutate(feature_ids = map(
    data,
    ~ pull(.x, feature_id)
  )) %>%
  select(-data)

# .x <- results_de_real_all_gene$de[[1]]
# .y <- mappability_mean_abundance_feature_ids$feature_ids[[1]]
# 
# h.de$calc_spearman(est = .x, true = result_qpcr, feature_ids = .y)

mappability_abundance_metrics_real_all_gene <- results_de_real_all_gene$de %>%
  enframe(value = "data") %>%
  filter(!grepl("sleuth", name)) %>%
  crossing(filter(mappability_mean_abundance_feature_ids, !is.na(bin.y))) %>%
  mutate(spearman = map2_dbl(
    data,
    feature_ids,
    ~ h.de$calc_spearman(est = .x, true = result_qpcr, feature_ids = .y)
  )) %>%
  mutate(nrmse = map2_dbl(
    data,
    feature_ids,
    ~ h.de$calc_nrmse(est = .x, true = result_qpcr, feature_ids = .y)
  )) %>%
  select(name, bin.x, bin.y, spearman, nrmse)
```
## Plot
```{r}
.data <- mappability_abundance_metrics_real_all_gene %>%
  pivot_longer(-(1:3), names_to = "metric") %>%
  utl$decorate_metrics() %>%
  utl$filter_by_combination("main") %>%
  filter(!is.na(bin.x)) %>%
  mutate(label_bin.y = factor(paste0("True abundance: ", bin.y))) %>%
  group_by(metric) %>%
  nest
  
# .data <- mappability_abundance_metrics_real_all_gene %>%
#   unnest(metrics) %>%
#   utl$decorate_metrics() %>%
#   utl$filter_by_combination("main") %>%
#   mutate(bin.x = factor(bin.x)) %>%
#   mutate(bin.y = factor(bin.y)) %>%
#   mutate(label_bin.y = factor(paste0("True abundance: ", bin.y))) %>%
#   group_by(dataset, metric) %>%
#   nest %>%
#   ungroup(dataset)

.draw1 <- function(data_, metric) {
  g <- utl$plot_bar(
    data = data_,
    group = "abbr",
    facet.by = c("bin.y", "bin.x"),
    tilt = TRUE,
    theme_ = NULL,
    list(
      x = "abbr",
      y = "value",
      xlab = "Bin of transcript mappability",
      ylab = "Bin of true transcript abundance"
    )
  )
}

.draw2 <- function(data_, metric) {
  g <- utl$plot_line(
    data = data_,
    group = "abbr",
    facet.by = c("label_bin.y"),
    tilt = TRUE,
    theme_ = NULL,
    list(
      x = "bin.x",
      y = "value",
      xlab = "Bin of transcript mappability",
      ylab = utl$to_lab(metric)
    )
  )
}

.data$plot1 <- .data %>%
  select(data, metric) %>%
  pmap(.draw1)

.data$plot2 <- .data %>%
  select(data, metric) %>%
  pmap(.draw2)

map2(
  .data$plot1,
  .data$metric,
  ~ .x %>% utl$save_plot(
    .y,
    "mappability_abundance_bar_de.pdf",
    width = 3.2 * 4.2,
    height = 3.2 * 4
  )
)

map2(
  .data$plot2,
  .data$metric,
  ~ .x %>% utl$save_plot(
    .y,
    "mappability_abundance_line_de.pdf",
    width = 3.2 * 4.2,
    height = 3.2 * 4
  )
)

g1 <- .data$plot2 %>%
  set_names(.data$metric)

rm(.data, .draw1, .draw2)
```

## Dist
```{r}
.mappabilities_abundances <- result_qpcr %>%
  select(feature_id, mean) %>%
  left_join(
    bind_rows(
      mappabilities %>% filter(name == "gencode" & feature_type == "gene") %>% .$data %>% .[[1]],
      mappabilities1 %>% filter(name == "gencode" & feature_type == "gene") %>% .$data %>% .[[1]]
    ),
    by = "feature_id"
  ) %>%
  filter(!is.na(feature_id))

.theme <- theme(
  text = element_text(size = 18),
  legend.position = "bottom",
  legend.title = element_blank(),
  legend.text = element_text(size = 16),
  strip.text.x = element_text(size = 16)
)

.g_main <- .mappabilities_abundances %>%
  filter(feature_id %in% pull(!!result_qpcr, "feature_id")) %>%
  mutate(log10mean = log10(mean * 100)) %>%
  mutate(biotype = utl$to_lab(biotype)) %>%
  filter(biotype %in% c("mRNA", "lncRNA")) %>%
  utl$plot_scatter_(
    group_ = "biotype",
    lims = NULL,
    tilt = FALSE,
    facet.by = NULL,
    theme_ = NULL,
    list(
      x = "mappability",
      y = "log10mean",
      xlab = "Mappability",
      ylab = "qPCR measurement (log10 %POLR2A)",
      alpha = 0.5
    )
  ) + scale_colour_manual(values = c(mRNA = "#00AFBB", lncRNA = "#E7B800")) + scale_fill_manual(values = c(mRNA = "#00AFBB", lncRNA = "#E7B800")) + guides(colour = guide_legend(override.aes = list(alpha = 1))) + .theme

.draw <- function(data_, x, labs) {
  .theme <- theme(
    text = element_text(size = 18),
    legend.position = "bottom",
    legend.title = element_blank(),
    legend.text = element_text(size = 18)
  )

  g <- data_ %>%
    mutate(log10mean = log10(mean * 100)) %>%
    mutate(biotype = utl$to_lab(biotype)) %>%
    filter(biotype %in% c("mRNA", "lncRNA")) %>%
    ggpubr::gghistogram(
      x = x,
      bins = 100,
      rug = FALSE,
      color = "biotype",
      fill = "biotype",
      alpha = 0.125,
      xlab = labs[[1]],
      ylab = labs[[2]],
      add = "mean"
      # facet.by = "sample"
    )
  g <- g + .theme
  g <-
    g + scale_colour_manual(values = c(mRNA = "#00AFBB", lncRNA = "#E7B800"))
  g <-
    g + scale_fill_manual(values = c(mRNA = "#00AFBB", lncRNA = "#E7B800"))

  g
}

.g_top <- .mappabilities_abundances %>%
  .draw("mappability", c("Mappability", "Frequency"))

.g_right <- .mappabilities_abundances %>%
  .draw("log10mean", c("qPCR measurement (log10 %POLR2A)", "Frequency"))

.legend <- ggpubr::get_legend(.g_top)

.arrange <- function(main, top, right) {
  .theme1 <- theme(
    text = element_text(size = 18),
    axis.text.x = element_blank(),
    axis.text.y = element_text(size = 16),
    plot.margin= unit(c(1, 0, 1, 0), "lines"),
    legend.title = element_blank(),
  )
  .theme2 <- theme(
    text = element_text(size = 18),
    axis.text.x = element_text(size = 16),
    axis.text.y = element_blank(),
    axis.title.y = element_blank(),
    strip.text = element_blank(),
    plot.margin = unit(c(0, 0, 0, 0), "lines"),
    legend.title = element_blank(),
  )

  .top <- top + theme_bw() + .theme1 + ggpubr::rremove("legend") + ggpubr::rremove("x.text") + ggpubr::rremove("x.ticks") + ggpubr::rremove("xlab")
  .right <- right + theme_bw() + .theme2 + ggpubr::rotate() + ggpubr::rremove("legend") + ggpubr::rremove("y.text") + ggpubr::rremove("ylab") + ggpubr::rremove("y.ticks")
  .main <- main + theme(plot.margin= unit(c(0, 0, 0, 0), "lines"))
  g <- ggpubr::ggarrange(
    plotlist = list(.top, NULL, .main, .right),
    ncol = 2, nrow = 2, align = "hv", hjust = 2, vjust = 3,
    widths = c(3, 1), heights = c(1, 3),
    common.legend = FALSE, legend = "none"
  )

  g
}

g2 <- .arrange(.g_main, .g_top, .g_right) + theme(plot.background = element_rect(fill = "white"))
g2 <- g2 %>% ggpubr::annotate_figure(bottom = .legend)

utl$save_plot(
    g1,
    "mappability_abundance",
    "distribution_biotype.pdf",
    width = 6.4 * 2.4,
    height = 4.8 * 2
  )

```


```{r}
.add_lab <- function(g, text) {
  (g + theme(plot.margin = unit(c(2, 1.5, 0.5, 0.5), "lines"))) %>%
    ggpubr::annotate_figure(fig.lab = text,
                            fig.lab.size = 21,
                            fig.lab.face = "bold")
}

g <- ggpubr::ggarrange(
  g1$spearman %>% .add_lab("(A)"),
  g2 %>% .add_lab("(B)"),
  ncol = 1,
  nrow = 2,
  align = "hv",
  heights = c(1, 2),
  common.legend = FALSE,
  legend = "bottom"
)

g %>% utl$save_plot("spearman_dist", "arranged_de.pdf", width = 3.2 * 4.2, height = 3.2 * 6)
```


### 2.1 Calculate metrics (vs. qPCR validation data)

```{r}
.metric_est_true <- function(est, true, method, feature_ids = c()) {
  if (length(feature_ids) < 1) feature_ids <- true$feature_id

  joined_ <- true %>% left_join(est, by = c("feature_id")) %>%
    filter(feature_id %in% feature_ids)

  if (method == "spearman") metric <- cor(joined_$logfc.x, joined_$logfc.y, method = "spearman", use = "complete.obs")
  if (method == "nrmse") metric <- nrmse(joined_$logfc.x, joined_$logfc.y, na.rm = TRUE)

  return(metric)
}

# NOTE: Ommmiting sleuth (because do not output log2fc value)
# TBC: Use Sleuth Wald test (beta stastics)?
.keep <-
  names(results_de_real_all_gene$de) %>% grepl("sleuth", .) %>% !.

.spearman <-
  results_de_real_all_gene$de[.keep] %>%
  sapply(.metric_est_true, result_qpcr, method = "spearman") %>%
  enframe %>%
  mutate(metric = "spearman")

.nrmse <- results_de_real_all_gene$de[.keep] %>%
  sapply(.metric_est_true, result_qpcr, method = "nrmse") %>%
  enframe %>%
  mutate(metric = "nrmse")

metrics_real_all_gene <- bind_rows(
  .spearman,
  .nrmse
) %>%
  utl$decorate_metrics() %>%
  utl$split_tibble("annotation")

.keep <- names(results_de_sim_all_gene$de) %>% grepl("sleuth", .) %>% !.

.spearman <- results_de_sim_all_gene$de[.keep] %>%
  sapply(.metric_est_true, result_qpcr, method = "spearman") %>% enframe %>%
  mutate(metric = "spearman")

.nrmse <- results_de_sim_all_gene$de[.keep] %>%
  sapply(.metric_est_true, result_qpcr, method = "nrmse") %>% enframe %>%
  mutate(metric = "nrmse")

metrics_sim_all_gene <-  bind_rows(
  .spearman,
  .nrmse
) %>%
  utl$decorate_metrics() %>%
  utl$split_tibble("annotation")
```

### 2.5 Draw correlations & NRMSE of log2CPM

```{r}
.dat <- metrics_real_all_gene %>%
  do.call(rbind, .) %>%
  utl$filter_by_combination("main") %>%
  mutate(annotation = utl$to_factor(annotation))

.draw <- function(df, metric_) {
  .title <- c(spearman = "Spearman's rho", nrmse = "NRMSE")[metric_]
  .palette <- df %>% distinct(abbr, tool) %>% select(abbr, tool) %>% deframe %>% sapply(utl$to_color)

  g <- df %>%
    filter(metric == metric_) %>%
    ggdotchart(
      x = "abbr", y = "value",
      color = "abbr",
      x.text.col = "abbr",
      palette = .palette,
      sorting = "desc",
      dot.size = 2,
      ylab = .title,
      y.text.col = TRUE,
      rotate = TRUE,
      ggtheme = theme_bw()
    )
  g <- facet(g + theme_cleveland() + theme(legend.position = "bottom", legend.title = element_blank()), facet.by = c("annotation"))

  return(g)
}

g <- .dat %>% .draw("spearman")
.output_path <- here::here(output_dir, paste(PREFIX, "spearman.all", "dotchart.pdf", sep = "_"))
ggsave(file = .output_path, plot = g, dpi = 300, width = 6.4 * 2, height = 3.2 * 2)

g <- .dat %>% .draw("nrmse")
.output_path <- here::here(output_dir, paste(PREFIX, "rnmse.all", "dotchart.pdf", sep = "_"))
ggsave(file = .output_path, plot = g, dpi = 300, width = 6.4 * 2, height = 3.2 * 2)
```

## 3. Mock comparison

### 3.1 Fig. 6 (B) Draw boxplot (MAQCA vs MAQCA)

```{r}
dummy1 <- crossing(
  name = results_de_real_real$sig %>% names,
  biotype = c('all', 'mrna', 'lncrna', 'other')
)

dat1_ <- results_de_real_real$sig %>%
  enframe %>%
  mutate(value = map(value,
                     ~ utl$set_biotype(.x, annotations$gencode_refseq))
  )

counts_real_ <- bind_rows(
  dat1_ %>%
    unnest(value) %>%
    group_by(name, biotype) %>% nest,
  dat1_ %>% unnest(value) %>%
    mutate(biotype = "all") %>%
    group_by(name, biotype) %>% nest
) %>%
  filter(!is.na(biotype)) %>%
  mutate(value = map(data, nrow) %>% unlist) %>%
  select(-data) %>%
  right_join(dummy1, by = c("name", "biotype")) %>%
  mutate(value = ifelse(is.na(value), 0, value)) %>%
  mutate(replicate = strsplit(name, "\\.") %>% sapply("[", 1) %>% as.numeric) %>%
  mutate(comparison = "real") %>%
  utl$decorate_metrics() %>%
  mutate(annotation = utl$to_factor(annotation)) %>%
  utl$filter_combination("main")

dummy2 <- crossing(
  name = results_de_real_mock$sig %>% names,
  biotype = c('all', 'mrna', 'lncrna', 'other')
)

dat2_ <- results_de_real_mock$sig %>%
  enframe %>%
  mutate(value = map(value,
                     ~ utl$set_biotype(.x, annotations$gencode_refseq))
  )

counts_mock_ <- bind_rows(
  dat2_ %>%
    unnest(value) %>%
    group_by(name, biotype) %>% nest,
  dat2_ %>% unnest(value) %>%
    mutate(biotype = "all") %>%
    group_by(name, biotype) %>% nest
) %>%
  filter(!is.na(biotype)) %>%
  mutate(value = map(data, nrow) %>% unlist) %>%
  select(-data) %>%
  right_join(dummy2, by = c("name", "biotype")) %>%
  mutate(value = ifelse(is.na(value), 0, value)) %>%
  mutate(replicate = strsplit(name, "\\.") %>% sapply("[", 1) %>% as.numeric) %>%
  mutate(comparison = "mock") %>%
  utl$decorate_metrics() %>%
  mutate(annotation = utl$to_factor(annotation)) %>%
  utl$filter_combination("main")

subplot_ <- function(df, title_ = NA) {
  palette_ <- df %>%
    distinct(abbr, tool) %>%
    deframe %>% sapply(utl$to_color)

  g <- df %>%
    ggboxplot(
      x = "abbr", y = "value",
      color = "abbr", palette = palette_,
      title = title_, rotate = TRUE
    )
  g <- g + xlab("") + ylab("Number of DEs")
  g <- facet(g + theme_bw() + theme(legend.position = "bottom", legend.title = element_blank()), facet.by = c("annotation"))

  return(g)
}

plot_ <- function(df1, df2, lab_add) {
  g1 <- df1 %>%
    subplot_("MAQCA vs. MAQCB")

  g2 <- df2 %>%
    subplot_("MAQCA vs. MAQCA")

  g <- ggarrange(
    g1,
    g2,
    ncol = 1, nrow = 2,
    legend = "bottom", common.legend = TRUE
  )

  g <- g %>% annotate_figure(
    bottom = text_grob(lab_add, size = 16)
  )
}

plots_ <- pmap(
  list(
    df1 = counts_real_ %>%
      arrange(biotype, name) %>%
      group_by(biotype) %>%
      nest %>% .$data,
    df2 = counts_mock_ %>%
      arrange(biotype, name) %>%
      group_by(biotype) %>%
      nest %>% .$data,
    lab_add = paste0("biotype = ", counts_real_ %>%
                       arrange(biotype, name) %>%
                       group_by(biotype) %>%
                       nest %>% .$biotype %>% utl$to_factor())
  ),
  plot_
)

map2(
  plots_,
  counts_real_ %>%
    arrange(biotype, name) %>%
    group_by(biotype) %>%
    nest %>% .$biotype,
  ~ utl$save_plot(.x, .y, "efdr_box.pdf")
)

path_ <- here::here(output_dir, paste(PREFIX, "real.mock_num_de.tsv", sep = "_"))
bind_rows(
  counts_real_ %>% select(1:5),
  counts_mock_ %>% select(1:5)
) %>% write_tsv(path_)
```

### 3.2 Fig. SXX Draw boxplot (MAQCB vs MAQCB)

```{r}
counts_real_ <- counts_real_

dummy2 <- crossing(
  name = results_de_real_mock_maqcb$sig %>% names,
  biotype = c('all', 'mrna', 'lncrna', 'other')
)

dat2_ <- results_de_real_mock_maqcb$sig %>%
  enframe %>%
  mutate(value = map(value,
                     ~ utl$set_biotype(.x, annotations$gencode_refseq))
  )

counts_mock_ <- bind_rows(
  dat2_ %>%
    unnest(value) %>%
    group_by(name, biotype) %>% nest,
  dat2_ %>% unnest(value) %>%
    mutate(biotype = "all") %>%
    group_by(name, biotype) %>% nest
) %>%
  filter(!is.na(biotype)) %>%
  mutate(value = map(data, nrow) %>% unlist) %>%
  select(-data) %>%
  right_join(dummy2, by = c("name", "biotype")) %>%
  mutate(value = ifelse(is.na(value), 0, value)) %>%
  mutate(replicate = strsplit(name, "\\.") %>% sapply("[", 1) %>% as.numeric) %>%
  mutate(comparison = "mock") %>%
  utl$decorate_metrics() %>%
  mutate(annotation = utl$to_factor(annotation)) %>%
  utl$filter_combination("main")

subplot_ <- function(df, title_ = NA) {
  palette_ <- df %>%
    distinct(abbr, tool) %>%
    deframe %>% sapply(utl$to_color)

  g <- df %>%
    ggboxplot(
      x = "abbr", y = "value",
      color = "abbr", palette = palette_,
      title = title_, rotate = TRUE
    )
  g <- g + xlab("") + ylab("Number of DEs")
  g <- facet(g + theme_bw() + theme(legend.position = "bottom", legend.title = element_blank()), facet.by = c("annotation"))

  return(g)
}

plot_ <- function(df2, lab_add) {
  g2 <- df2 %>%
    subplot_("MAQCB vs. MAQCB")

  g <- ggarrange(
    g2,
    ncol = 1, nrow = 1,
    legend = "bottom", common.legend = TRUE
  )

  g <- g %>% annotate_figure(
    bottom = text_grob(lab_add, size = 16)
  )
}

plots_ <- pmap(
  list(
    df2 = counts_mock_ %>%
      arrange(biotype, name) %>%
      group_by(biotype) %>%
      nest %>% .$data,
    lab_add = paste0("biotype = ", counts_real_ %>%
                       arrange(biotype, name) %>%
                       group_by(biotype) %>%
                       nest %>% .$biotype %>% utl$to_factor())
  ),
  plot_
)

map2(
  plots_,
  counts_real_ %>%
    arrange(biotype, name) %>%
    group_by(biotype) %>%
    nest %>% .$biotype,
  ~ utl$save_plot(.x, .y, "efdr_box_maqcb.pdf", height = 3.2)
)

path_ <- here::here(output_dir, paste(PREFIX, "real.mock_num_de_maqcb.tsv", sep = "_"))
bind_rows(
  counts_real_ %>% select(1:5),
  counts_mock_ %>% select(1:5)
) %>% write_tsv(path_)
```

## 4. Comparison between real data and simulation data

### 4.1 Fig. SXX Draw dotchart

```{r}
metrics_ <- metrics_sim_all_gene %>%
  do.call("rbind", .) %>%
  mutate(annotation = utl$to_factor(annotation))

palette_ <- metrics_ %>% distinct(abbr, tool) %>% deframe %>% sapply(utl$to_color)

plot_ <- function(df, metric_) {
  title_ <- c(spearman = "Spearman's rho", nrmse = "NRMSE")[metric_]
  g <- df %>%
    filter(metric == metric_) %>%
    ggdotchart(
      x = "abbr", y = "value",
      color = "abbr",
      x.text.col	 = "abbr",
      palette = palette_,
      sorting = "desc",
      dot.size = 2,
      ylab = title_,
      y.text.col = TRUE,
      rotate = TRUE,
      ggtheme = theme_bw()
    )
  g <- facet(g + theme_cleveland() + theme(legend.position = "bottom", legend.title = element_blank()), facet.by = c("annotation"))

  return(g)
}

g <- metrics_ %>% plot_("spearman")
output_path_ <- here::here(output_dir, paste("de_sim", "spearman.all", "dotchart.pdf", sep = "_"))
ggsave(file = output_path_, plot = g, dpi = 300, width = 6.4 * 2, height = 3.2 * 2)

g <- metrics_ %>% plot_("nrmse")
output_path_ <- here::here(output_dir, paste("de_sim", "rnmse.all", "dotchart.pdf", sep = "_"))
ggsave(file = output_path_, plot = g, dpi = 300, width = 6.4 * 2, height = 3.2 * 2)
```

## 5. Overlap of DEs (main case only)

### 5.1 Obtain sets

```{r}
intersect_ <- function(x, y, z) {
  intersect_ <- x %>% intersect(
    annotations[[y]] %>% filter(biotype == z) %>% .$transcript_id
  )

  return(intersect_)
}

annotations_used_ <- names(results_de_real_real$sig) %>% lapply(utl$to_combination) %>% sapply(utl$annotation_used)

mintersect_ <- function(biotype_) {
  mapply(
    partial(intersect_, z = biotype_),
    results_de_real_real$sig,
    annotations_used_,
    SIMPLIFY = FALSE
  )
}

results_de_real_real$sig %>% head

de_sig_biotypes <- list(
  mrna = mintersect_("mrna"),
  lncrna = mintersect_("lncrna"),
  other = mintersect_("other")
)
```

### 5.2 Obtain pairwise intersection

```{r}
dat_ <- results_de_real_real$sig %>%
  enframe %>%
  mutate(value = map(value,
                     ~ utl$set_biotype(.x, annotations$gencode_refseq))
  )

biotype_feature_ids_sig <- bind_rows(
  dat_ %>%
    unnest(value) %>%
    group_by(name, biotype) %>% nest,
  dat_ %>% unnest(value) %>%
    mutate(biotype = "all") %>%
    group_by(name, biotype) %>% nest
) %>%
  filter(!is.na(biotype)) %>%
  mutate(feature_ids = map(data, ~ pull(.x, feature_id))) %>%
  select(-data) %>%
  right_join(dummy1, by = c("name", "biotype")) %>%
  mutate(feature_ids = ifelse(is.na(feature_ids), c(), feature_ids)) %>%
  mutate(replicate = strsplit(name, "\\.") %>% sapply("[", 1) %>% as.numeric) %>%
  mutate(comparison = "real") %>%
  utl$decorate_metrics() %>%
  mutate(annotation = utl$to_factor(annotation)) %>%
  utl$filter_combination("main") %>%
  filter(grepl("00\\.", name))

biotype_feature_ids_sig <- biotype_feature_ids_sig %>%
  select(biotype, name, feature_ids) %>%
  group_by(biotype) %>%
  nest

n_intersect_ <- function(list_, rel = TRUE) {
  names(list_) <- names(list_) %>% strsplit("\\.") %>% sapply("[", 2) %>% utl$to_abbr(to_scenario = FALSE)

  n_ <- function(query, sets_) {
    n_ <- sets_ %>% lapply(intersect, query) %>% lapply(length)
    return(n_)
  }

  n_intersect <- list_ %>%
    lapply(n_, sets_ = list_)

  df_n_intersect <- n_intersect %>%
    unlist %>%
    matrix(ncol = length(n_intersect)) %>%
    data.frame

  rownames(df_n_intersect) <- colnames(df_n_intersect) <- names(n_intersect)

  abs2rel_ <- function(x) {
    # NOTE: The query itself always equal max
    denominator <- max(x)

    result <- x / denominator
    result[is.nan(result)] <- 0
    return(result)
  }

  if (rel) return(apply(df_n_intersect, 1, abs2rel_))
  return(df_n_intersect)
}

biotype_feature_ids_sig <- biotype_feature_ids_sig %>%
  mutate(intersect = map(data, deframe) %>% map(n_intersect_))
```

### 5.3 Fig. 6 (C) Draw pairwise intersection

```{r}
plots_ <- map(
  biotype_feature_ids_sig$intersect,
  utl$plot_intersect
) %>% map2(
  .,
  biotype_feature_ids_sig$biotype,
  ~ annotate_figure(.x, bottom = text_grob(paste0("biotype = ", utl$to_factor(.y)), size = 16))
)

map2(
  plots_,
  biotype_feature_ids_sig$biotype %>% tolower,
  ~ utl$save_plot(.x, .y, "intersections.pdf")
)
```
