---
title: "Generation for simulation count data from source real data"
output: html_notebook
editor_options:
  chunk_output_type: console
---

## Set enviroment & requirements

```{r, results='hide'}
options(stringAsFactors = FALSE)

library(readr)
library(data.table)

library(corrplot)
library(patchwork)
library(edgeR)

library(tidyverse)

utl <- new.env(); source(here::here("scripts/utils/base.R"), utl); source(here::here("scripts/utils/plot.R"), utl)
h.quant <- new.env(); source(here::here("scripts/helpers/quant_helper.R"), h.quant)

task_name <- 'results/sim_cntmat/gencode'
dir.create(here::here(task_name), showWarnings = FALSE)
```

## Set global parameters

```{r}
PSUEDO_COUNT <- 1e-2
SEED <- 12345

LIB_SIZE <- 8e7
N_REP <- 24
```

## Load gene annotation

```{r}
load_gtf <- function(from = root_dir) {
  path <- here::here('shared/assets/references/grch38/annotations/gencode/gencode.v31.annotation.gtf')
  cols <- c('feature', 'gene_id', 'gene_name', 'gene_type', 'transcript_id', 'transcript_name', 'transcript_type')
  gtf <- rtracklayer::readGFF(path, version = 2L, tags = cols, filter = list(type = c('transcript')))

  path <- here::here('shared/assets/references/grch38/annotations/gencode/gencode.v31.long_noncoding_RNAs.gtf')
  cols <- c('gene_id', 'transcript_id')
  gtf_lncrna <- rtracklayer::readGFF(path, version = 2L, tags = cols, filter = list(type = c('transcript')))

  gtf$biotype <- 'Other'
  gtf$biotype[gtf$transcript_id %in% gtf_lncrna$transcript_id[!is.na(gtf_lncrna$transcript_id)]] <- 'lncRNA'
  gtf$biotype[gtf$transcript_type == 'protein_coding'] <- 'mRNA'
  return(gtf)
}

gtf <- load_gtf()
```

## Create edgeR object

```{r}
perform_exact_test <- function(counts, groups = NULL, filter = TRUE) {
  n_rep <- ncol(counts) / 2
  if (length(groups) < 1) groups <- c(rep('ctrl', n_rep), rep('case', n_rep))
  colnames(counts) <- paste0(groups, '_', seq(n_rep) - 1)

  if (filter) {
    keep <- rowSums(counts > 1) >= n_rep
    counts <- counts[keep, ]
  }

  d <- DGEList(counts = counts, group = groups)
  d <- calcNormFactors(d, method = 'TMM')
  d <- estimateCommonDisp(d)
  d <- estimateTagwiseDisp(d)
  d <- calcNormFactors(d)

  de <- exactTest(d, pair = c('ctrl', 'case'))

  edger <- list(
    d = d,
    de = de
  )

  return(edger)
}

remove_spikein <- function(mat, prefix = 'ERCC-') {
  keep <- !grepl(prefix, rownames(mat))
  return(mat[keep, ])
}

counts_src <- here::here('results/prequant_star_rsem/gencode/rsem.isoforms.result.matrix.tsv') %>% fread(sep = '\t') %>% column_to_rownames('V1') %>% as.matrix
counts_src <- remove_spikein(counts_src)
counts_src %>% data.frame %>% rownames_to_column('feature_id') %>% write_tsv(path = here::here(task_name, 'counts_src.txt'))
meta <- here::here('assets/datasets/PRJNA362835/meta/summary_intact_gt_2ng.txt') %>% fread(sep = '\t')

lookup_group <- function(mat, meta) {
  groups <- list()
  for (ra in meta$run_accession) {
    idx <- grep(ra, colnames(mat))
    groups[idx] <- meta[idx, 'run_alias']
  }
  groups <- unlist(groups, use.names=FALSE)
  names(groups) <- colnames(mat)
  return(groups)
}

lookup_group(counts_src, meta) %>% message

edger_src <- perform_exact_test(counts = counts_src, groups = c(rep('case', 24), rep('ctrl', 24)))
edger_src$d %>% cpm %>% data.frame %>% rownames_to_column('feature_id') %>% write_tsv(path = here::here(task_name, 'cpm_src.txt'))
edger_src$de %>% topTags(n = nrow(edger_src$de)) %>% data.frame %>% rownames_to_column('feature_id') %>% write_tsv(path = here::here(task_name, 'edger_src.txt'))
```

## Calculate simulation parameters from edgeR object

```{r}
mean_by_group <- function(mat, prefixes = c('ctrl', 'case')) {
  mean <- data.frame(
    ctrl = mat[, grepl(prefixes[1], colnames(mat))] %>% apply(1, mean),
    case = mat[, grepl(prefixes[2], colnames(mat))] %>% apply(1, mean))
  return(mean)
}

calc_sim_params <- function(obj, psuedo_count = PSUEDO_COUNT) {
  d <- obj$d
  de <- obj$de
  disp <- d$tagwise.dispersion
  cpm <- cpm(d, log = FALSE) + psuedo_count

  mean_cpm <- mean_by_group(cpm)

  # NOTE: Normalized CPM as fold change
  # fc <- 2^de$table$logFC
  fc <- mean_cpm$case / mean_cpm$ctrl

  params <- list(
    fid = rownames(cpm),
    lambda = mean_cpm$ctrl,
    disp = disp,
    fc = fc,
    mean_cpm = mean_cpm
  )

  return(params)
}

sim_params <- calc_sim_params(edger_src)
data.frame(
  feature_id = sim_params$fid,
  lambda = sim_params$lambda,
  disp = sim_params$disp,
  fc = sim_params$fc
  ) %>% write_tsv(path = here::here(task_name, 'params_sim.txt'))

'Number of DEs' %>% message
(sim_params$fc >= 2) %>% table %>% message
```

## Generate simulation counts matrix from NB distribution

```{r}
sim_counts <- function(p, lib_size, n_rep, n_grp = 2, seed = SEED, prefixes = c('ctrl', 'case')) {
  set.seed(seed)
  p <- sim_params
  n_feature <- length(p$fid)
  groups <- c(rep(prefixes[1], n_rep), rep(prefixes[2], n_rep))

  base_mu <- (p$lambda / sum(p$lambda)) * lib_size
  mu_ctrl <- expandAsMatrix(base_mu, dim = c(n_feature, n_rep))
  mu_case <- expandAsMatrix(base_mu * p$fc, dim = c(n_feature, n_rep))
  mu <- cbind(mu_ctrl, mu_case)

  counts <- rnbinom(n_feature * n_rep * n_grp, mu = mu, size = 1 / p$disp) %>% matrix(nrow = n_feature, ncol = n_rep * n_grp)
  rownames(counts) <- p$fid
  colnames(counts) <- paste0(groups, '_', seq(n_rep) - 1)
  return(counts)
}

counts_sim <- sim_counts(sim_params, lib_size = LIB_SIZE, n_rep = N_REP)
counts_sim %>% data.frame %>% rownames_to_column('feature_id') %>% write_tsv(path = here::here(task_name, 'counts_sim.txt'))
edger_sim <- perform_exact_test(counts = counts_sim, filter = FALSE)
edger_sim$d %>% cpm %>% data.frame %>% rownames_to_column('feature_id') %>% write_tsv(path = here::here(task_name, 'cpm_sim.txt'))
edger_sim$de %>% topTags(n = nrow(edger_sim$de)) %>% data.frame %>% rownames_to_column('feature_id') %>% write_tsv(path = here::here(task_name, 'edger_sim.txt'))
sim_params_sim <- calc_sim_params(edger_sim)
```

## Compare source data and simulated data

```{r}
keep <- rowSums(counts_src > 1) >= 24
counts_src <- counts_src[keep,]

colnames(counts_src) <- NULL

"Library sizes" %>% message
counts_src %>% colSums %>% message
counts_sim %>% colSums %>% message

"Summary" %>% message
counts_src %>% summary %>% message
counts_sim %>% summary %>% message

"Correlations" %>% message
counts_src %>% cor %>% corrplot
counts_src %>% cor %>% data.frame
counts_sim %>% cor %>% corrplot
counts_sim %>% cor %>% data.frame

mean_cpm_src <- sim_params$mean_cpm

mean_cpm_sim <-
  (cpm(edger_sim$d, log = FALSE) + PSUEDO_COUNT) %>%
    mean_by_group(prefixes = c('ctrl', 'case'))

mean_cpm_src <-
  mean_cpm_src %>%
  rownames_to_column('feature_id') %>%
  gather(key = group, value = cpm, -1) %>%
  arrange(feature_id, group)

mean_cpm_sim <-
  mean_cpm_sim %>%
  rownames_to_column('feature_id') %>%
  gather(key = group, value = cpm, -1) %>%
  arrange(feature_id, group)

mean_cpm_src <- mean_cpm_src %>% mutate(log2cpm = log2(cpm))

mean_cpm_sim <- mean_cpm_sim %>% mutate(log2cpm = log2(cpm))

"Pearson's and Spearman's corr" %>% message
cor(mean_cpm_src$log2cpm, mean_cpm_sim$log2cpm, method = 'pearson') %>% message
cor(mean_cpm_src$log2cpm, mean_cpm_sim$log2cpm, method = 'spearman') %>% message

plot_box <- function(df, xcol, ycol, stitle = NULl) {
  g <- ggplot(df, aes_string(x = xcol, y = ycol)) +
    geom_boxplot(notch = TRUE) +
    labs(title = 'Dist. log2(raw counts)', subtitle = stitle) +
    theme_light()
  return(g)
}

g1 <-
  log2(counts_src + 1) %>% data.frame %>% rownames_to_column('feature_id') %>% gather(key = sample, value = count,-1) %>% plot_box('sample', 'count', 'Source')
g2 <-
  log2(counts_sim + 1) %>% data.frame %>% rownames_to_column('feature_id') %>% gather(key = sample, value = count,-1) %>% plot_box('sample', 'count', 'Simulation')

wrap_plots(g1, g2) %>% plot

plot_dist <-
  function(df,
           xcol,
           gcol = NULL,
           title = NULL,
           stitle = NULL) {
    g <- df %>% ggplot(aes_string(x = xcol))
    g <-
      g + geom_histogram(
        aes_string(color = gcol, fill = gcol),
        position = 'identity',
        bins = 30,
        alpha = 0.3
      )
    g <- g + labs(title = title, subtitle = stitle)
    g <- g + theme_light()
    return(g)
  }

g3 <-
  mean_cpm_src %>% filter(group == 'ctrl') %>% left_join(gtf, by = c('feature_id' = 'transcript_id')) %>% plot_dist('log2cpm', 'biotype', NULL, 'Source (ctrl)')
g4 <-
  mean_cpm_src %>% filter(group == 'case') %>% left_join(gtf, by = c('feature_id' = 'transcript_id')) %>% plot_dist('log2cpm', 'biotype', NULL, 'Source (case)')
g5 <-
  mean_cpm_sim %>% filter(group == 'ctrl') %>% left_join(gtf, by = c('feature_id' = 'transcript_id')) %>% plot_dist('log2cpm', 'biotype', NULL, 'Simulation (ctrl)')
g6 <-
  mean_cpm_sim %>% filter(group == 'case') %>% left_join(gtf, by = c('feature_id' = 'transcript_id')) %>% plot_dist('log2cpm', 'biotype', NULL, 'Simulation (case)')

(
  wrap_plots(g3, g4, g5, g6, ncol = 2, byrow = FALSE) + plot_annotation(title = 'Distribution of log2CPM')
) %>% plot

plot_ma <- function(df, xcol, ycol) {
  g <- df %>% ggplot(aes_string(x = xcol, y = ycol)) +
    geom_point() +
    xlab('Average logCPM') +
    ylab('logFC') +
    theme_light()
  return(g)
}

'plot BCV and MA (Source)' %>% message
edger_src$d %>% plotBCV
edger_src$de %>% plotSmear(cex = 0.3)

'plot BCV and MA (Simulation)' %>% message
edger_sim$d %>% plotBCV
edger_sim$de %>% plotSmear(cex = 0.3)

# UNUSED: Plot manually
# data.frame(log2cpm = log2(rowMeans(sim_params$mean_cpm)),
#            log2fc = log2(sim_params$fc)) %>% plot_ma('log2cpm', 'log2fc') %>% plot

edger_src$de$table %>% data.frame %>% summary %>% message
data.frame(fc = sim_params$fc) %>% log2 %>% summary %>% message

g7 <-
  edger_src$de$table %>% data.frame %>% rownames_to_column('feature_id') %>% left_join(gtf, by = c('feature_id' = 'transcript_id')) %>% plot_dist('logFC', 'biotype', NULL, 'Source-edgeR')
g8 <-
  edger_sim$de$table %>% data.frame %>% rownames_to_column('feature_id') %>% left_join(gtf, by = c('feature_id' = 'transcript_id')) %>% plot_dist('logFC', 'biotype', NULL, 'Simulation-edgeR')

(wrap_plots(g7, g8) + plot_annotation(title = 'Distribution of log2FC')) %>% plot

```

## Work
```{r}
path_annotation <- here::here("shared/assets/references/grch38/annotations/gencode/gencode.v31.annotation.gtf")

transcript_lengths <- path_annotation %>%
  utl$load_gtf(types = c("exon")) %>%
  group_by(transcript_id) %>%
  summarise(length = sum(end - start + 1)) %>%
  data.frame

mean_cpm_sim_agg <- mean_cpm_sim %>%
  group_by(feature_id) %>%
  summarize(log10cpm_mean = log10(mean(cpm)))

mean_tpm_sim_agg <- mean_cpm_sim %>%
  group_by(feature_id) %>%
  summarize(cpm_mean = mean(cpm)) %>%
  h.quant$to_tpm(., transcript_lengths) %>%
  dplyr::rename(tpm_mean = cpm_mean) %>%
  mutate(log10tpm_mean = log10(tpm_mean))

g1 <- mean_cpm_sim_agg %>%
  left_join(gtf, by = c('feature_id' = 'transcript_id')) %>%
  # filter(log10cpm_mean < 1) %>%
  plot_dist('log10cpm_mean', 'biotype', NULL, 'Simulation (all)')

g2 <- mean_tpm_sim_agg %>%
  left_join(gtf, by = c('feature_id' = 'transcript_id')) %>%
  plot_dist('log10tpm_mean', 'biotype', NULL, 'Simulation (all)')

g3 <- mean_cpm_sim_agg %>%
  left_join(gtf, by = c('feature_id' = 'transcript_id')) %>%
  ggpubr::ggboxplot(x = "biotype", y = "log10cpm_mean")

g4 <- mean_tpm_sim_agg %>%
  left_join(gtf, by = c('feature_id' = 'transcript_id')) %>%
  ggpubr::ggboxplot(x = "biotype", y = "log10tpm_mean")

ggpubr::ggarrange(g1, g2)

ggpubr::ggarrange(g3, g4)

mean_cpm_sim$cpm %>% sum
mean_cpm_sim_agg$log10cpm_mean %>% sum
mean_tpm_sim_agg$log10tpm_mean %>% sum

.groups <- counts_sim %>%
  colnames %>%
  strsplit("_") %>%
  sapply("[", 1) %>%
  data.frame(cond = .)

dds1 <-
  DESeq2::DESeqDataSetFromMatrix(countData = apply(counts_src, c(1, 2), ceiling),
                                 colData = .groups,
                                 design = ~ cond)

dds2 <-
  DESeq2::DESeqDataSetFromMatrix(countData = counts_sim,
                                 colData = .groups,
                                 design = ~ cond)

countsimQCReport(
  ddsList = list(Source = dds1, Simulated = dds2),
  outputFile = "countsimReport2.html",
  outputDir = here::here(),
  description = "This is a comparison of three count data sets."
)


```
